### ISSUE TITLE
[Feature request] ggml as an NX backend

### ISSUE THREAD

**Issue by kipcole9** [elixir-nx/nx#1245]
It would be wonderful if [ggml](https://github.com/ggerganov/ggml) could be an Nx backend - especially for MacOS / ARM users. I appreciate that's a big ask. Hopefully there is someone motivated to try an implementation and has the requisite skill.

**Comment by polvalente** [comment]
I'm not sure we should have this open as an issue, though I see value in this being _somewhere around here_.
@josevalim WDYT about this being opened as a discussion instead?

Main reasoning for this not being an issue: a backend it's more like a client library that depends on Nx than a feature of Nx itself (see EXLA and Torchx, for example)

**Comment by kipcole9** [comment]
Totally reasonable - I just didn't know where else to surface it.

**Comment by polvalente** [comment]
Yes. We had a few similar things come up in the Slack channel, so this is why perhaps we should open a new channel on github as well. Also would help with visibility I think.

**Comment by josevalim** [comment]
The interesting thing about ggml is that the whole model is implemented as a single C++ program. We could provide Nx bindings for its underlying C operations, but then you introduce the back and forth between C and Elixir, and I don't think it would be much different than Torchx. So, in my mind, I think the best way to integrate with ggml is by integrating directly with the models, this is what Go and Rust folks are doing: https://github.com/ggerganov/ggml#resources

Another potential source of inspiration here is Ortex: https://github.com/relaypro-open/ortex - it defines a storage-only Nx backend for the tensors and a mechanism to run serving on top of existing models. I can see the GGML integration going pretty much along the same lines.

**Comment by josevalim** [comment]
Closing this for now. I think ggml would be awesome with the constraints above if someone wants to explore them.
